### PR TITLE
feat(agent): add streaming model interceptor for per-chunk LLM output

### PR DESCRIPTION
**Pull request by Jackie9991**
### Describe what this PR does / why we need it

新增 `StreamingModelInterceptor` —— 面向 `ReactAgent` 流式调用的扩展点，让用户在 `agent.stream(...)` 过程中按 `ChatResponse` chunk 粒度进行拦截。

当前框架已有四种拦截层：`Flux<NodeOutput>.doOnNext`（节点级）、`AgentHook`/`ModelHook`（模型调用边界）、`ModelInterceptor`（整次调用）、Spring AI `Advisor`（ChatClient 层），但**没有任何一个提供干净的、官方支持的 chunk 级回调**。要在今天实现按 chunk 处理，只能在 `ModelInterceptor` 内部把 `ModelResponse.getMessage()` 强转成 `Flux<ChatResponse>` 再 `.map()`，写法别扭、且把流式与阻塞语义混在一起。

本 PR 把这个空缺补上，让 chunk 级需求（逐 chunk 日志/指标、流式敏感词过滤、SSE 转换、内容审核、实时埋点等）成为一等公民 API：

```java
ReactAgent agent = ReactAgent.builder()
        .name("my-agent")
        .model(chatModel)
        .streamingInterceptors(myInterceptor)
        .build();
```

### Does this pull request fix one issue?

NONE

### Describe how you did it

**1. 新增 `StreamingModelInterceptor` 接口**（`spring-ai-alibaba-agent-framework/.../interceptor/StreamingModelInterceptor.java`），含四个 `default` 生命周期回调：

| 方法 | 触发时机 | 用途 |
|---|---|---|
| `beforeStreamCall(ModelRequest)` | 每次订阅前调用一次 | 在拦截器之间传递 context（返回的 request 作为后续回调的参数）。**不影响外发模型调用**——要改外发请求请用 `ModelInterceptor` |
| `onStreamChunk(ChatResponse, ModelRequest)` | 每个 chunk | 检视、转换或过滤 chunk（返回 `null` 即丢弃） |
| `afterStreamComplete(AssistantMessage, ModelRequest)` | 每次成功完成时调用一次 | 拿到**聚合后**的完整 AssistantMessage 用于日志/指标 |
| `onStreamError(Throwable, ModelRequest)` | 流出错时 | 错误日志 / 告警 |

**2. 链式实现** `InterceptorChain.applyStreamingInterceptors(...)` 把多个拦截器组合到 `Flux<ChatResponse>` 上。**注册顺序里第一个 = 最内层包装**——它最先看到模型原始 chunk，后面的拦截器看到的是上一个返回的 chunk。这个顺序与同步的 `chainModelInterceptors` / `chainToolInterceptors` 相反（那两个是 first = outermost），因为 chunk 级拦截器的核心价值是观察模型原始输出，源头优先更直觉。整条管道用 `Flux.defer(...)` 包裹，每次订阅都会跑一遍 `beforeStreamCall` 并拿到独立的聚合缓冲，避免 retry / 多订阅者的状态串扰与多线程 race。`onStreamChunk` 改用 `Flux.handle(...)`，用户返回 `null` 即可丢弃该 chunk（filter 语义）。

**3. 聚合语义修复** —— 之前的 `applyStreamingInterceptors` 把**最后一个 chunk** 的 `AssistantMessage` 传给 `afterStreamComplete`。对 DashScope / OpenAI 这类 delta 风格的 provider 而言，最后一个 chunk 只是末尾增量，并不是完整文本。修复后的链路改为对所有 chunk 累加 `chunk.getResult().getOutput().getText()`，再用完整文本构造一个 `AssistantMessage` 传出。

**4. 接线**（让拦截器真正被调用的"最后一公里"）：
- `Builder.java`：新增字段 `streamingInterceptors` + 两个 `streamingInterceptors(...)` 重载
- `ReactAgent.java`：读取 `builder.streamingInterceptors`，在已有的 model/tool 拦截器接线旁调用 `llmNode.setStreamingInterceptors(...)`
- `AgentLlmNode.java`：新增字段 + `setStreamingInterceptors(...)`；在流式 `baseHandler` 内，`enableReasoningLog` 之后、`ModelResponse.of(chatResponseFlux)` 之前调用 `InterceptorChain.applyStreamingInterceptors(...)` 把 Flux 包一层。挂在 base handler **内部**（相对 `ModelInterceptor` 是最内层），意味着 streaming 拦截器看到的是模型的原始 chunk；外层 `ModelInterceptor` 仍可对结果 Flux 做整次调用层面的处理（retry、fallback 等）。

**5. 测试重构**：
- 新增 `AbstractDashscopeChatModelTest` base —— 提供共享的 `protected ChatModel chatModel`，用环境变量驱动 OpenAI 兼容 DashScope 端点初始化；当 `AI_DASHSCOPE_API_KEY` 缺失时自动跳过。
- 新增 `StreamingModelInterceptorTest` —— 本扩展点的端到端测试（用 `QueryProduct1Tool` 强制走两轮流式：一轮决定调用工具 + 一轮基于工具结果作答）。

### Describe how to verify it

```bash
# 设置 DashScope 凭据（任意非空 AI_DASHSCOPE_API_KEY 即可启用受控测试）
export AI_DASHSCOPE_API_KEY=<your_dashscope_key>

# 跑新增的流式拦截器端到端测试
./mvnw -pl spring-ai-alibaba-agent-framework -am test \
  -Dtest='StreamingModelInterceptorTest#streamingModelInterceptor_shouldReceivePerChunkCallbacks'
```

预期运行输出（计数随模型详尽程度浮动）：

```
[interceptor] >>> beforeStreamCall round=1, messages=2, tools=[query_product_1]
[interceptor] chunk#1 toolCalls=[ToolCall[id=..., name=query_product_1, arguments=...]]
[interceptor] <<< afterStreamComplete round=1, aggregatedText=
[interceptor] >>> beforeStreamCall round=2, messages=4, tools=[query_product_1]
[interceptor] chunk#3 text=商品
[interceptor] chunk#4 text=1是编号
... (per-chunk delta text continues) ...
[interceptor] <<< afterStreamComplete round=2, aggregatedText=<完整最终回答>

===== Interceptor 计数 =====
beforeStreamCall   : 2
onStreamChunk      : 39       (≥ beforeStreamCall，随模型而异)
afterStreamComplete: 2
onStreamError      : 0
```

测试断言：`beforeStreamCall ≥ 2`、`chunkCount ≥ beforeCount`、`completeCount ≥ 2`、`errorCount == 0`，并且聚合文本非空（验证 chunk text 累加修复生效）。

### Special notes for reviews

- **兼容性**：纯加法改动。已有 `ModelInterceptor` / `ToolInterceptor` / `Hook` 链路完全不动；用户如仍想用 `ModelInterceptor` 自行包装流式响应也可以照旧。未改动任何已有公开类型的方法签名。
- **挂载点选择**：streaming 拦截器挂在 `baseHandler` **内部**（最内层），而不是 `chainedHandler.call()` 之后。原因：chunk 级拦截器应观察模型的原始输出；如果某个 `ModelInterceptor`（比如 retry）重新调用 handler，每次重试都会自然地重新走一遍 streaming 链，符合直觉。
- **`StreamingModelInterceptor` 没有 `extends Interceptor`** —— 故意保持独立，因为它语义上和现有 `Interceptor`（chunk 级 vs 整次调用）不同，注册入口也独立成 `streamingInterceptors(...)`，与 `interceptors(...)` 平级。如果后续想统一，挂到 `Interceptor` 体系下做 follow-up 也很干净。
- **`Builder.chatClient(...)` + Spring AI `StreamAroundAdvisor`** 仍是想直接复用 Spring AI advisor 生态的现有逃生通道。
- **`beforeStreamCall` 的 request 是 context-only**：返回值仅串到本拦截器与后续拦截器的 `onStreamChunk`/`afterStreamComplete`/`onStreamError` 回调，**不会改写实际发到模型的请求**（chunk Flux 在拦截器跑之前已经构造好）。要改外发 request 请用 `ModelInterceptor`。这一限制已在 `StreamingModelInterceptor` 的 Javadoc 里写明。
- **测试环境**：`StreamingModelInterceptorTest` 属于集成测试（真实调用 DashScope），由 `AI_DASHSCOPE_API_KEY` 控制；CI 不配 secret 时会被 `Assumptions.assumeTrue` 干净跳过。

### CODE REVIEW

**Comment by Jackie9991** [comment]
@Aias00 PTAL~

**Comment by Jackie9991** [comment]
@Aias00 done

**Comment by Jackie9991** [comment]
@Aias00 done

File: spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/interceptor/InterceptorChain.java [hunk 1]
```diff
@@ -94,7 +99,93 @@ public static ToolCallHandler chainToolInterceptors(
...
+			Flux<ChatResponse> flux,
+			ModelRequest request) {
+
+		if (interceptors == null || interceptors.isEmpty()) {
+			return flux;
+		}
+
+		return Flux.defer(() -> {
+			// Per-subscription: thread request through beforeStreamCall (first to last)
+			ModelRequest threaded = request;
```
Comment (Aias00) [inline spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/interceptor/InterceptorChain.java]: `Flux.defer(...)` makes the callbacks run per subscription, but each subscription still starts from the same `ModelRequest` object here. `ModelRequest` exposes mutable `context`, `messages`, and `options`, so a natural `beforeStreamCall` implementation that adds correlation IDs or tweaks options in-place will leak state across retries / multi-subscribe and can race under concurrent subscribers. If the goal is subscription-scoped state, this needs to start from a defensive copy (and ideally deep-copy the nested mutable fields too).

File: spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/interceptor/InterceptorChain.java [hunk 2]
```diff
@@ -94,7 +99,93 @@ public static ToolCallHandler chainToolInterceptors(
...
+			for (StreamingModelInterceptor interceptor : interceptors) {
+				threaded = interceptor.beforeStreamCall(threaded);
```
Comment (Aias00) [inline spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/interceptor/InterceptorChain.java]: Because `beforeStreamCall(...)` now runs inside the `Flux.defer` supplier, an exception thrown here happens before the `doOnError(...)` handlers below are attached. In that case `onStreamError(...)` is never invoked for any interceptor, which leaves a hole in the documented lifecycle. Could we either wrap the `beforeStreamCall` phase and fan out `onStreamError(...)` manually, or move it into the reactive chain so startup failures are observable through the same hook?

File: spring-ai-alibaba-agent-framework/src/test/java/com/alibaba/cloud/ai/graph/agent/interceptors/StreamingModelInterceptorTest.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+				.description("用于测试 StreamingModelInterceptor")
+				.instruction("请调用 query_product_1 工具查询商品信息，并基于返回结果用 500 字内说明这是什么商品，有什么特点等。")
+				.tools(List.of(QueryProduct1Tool.createCallback()))
+				.streamingInterceptors(counting)
+				.build();
+
+		try {
+			// Bounded wait so a hung upstream stream surfaces as a deterministic failure
+			// instead of blocking the test runner forever.
+			agent.stream("查询商品1并简要解读")
```
Comment (Aias00) [inline spring-ai-alibaba-agent-framework/src/test/java/com/alibaba/cloud/ai/graph/agent/interceptors/StreamingModelInterceptorTest.java:151]: This proves the live happy path, but it still does not pin down the risky semantics introduced in this PR: retry / multi-subscribe isolation, `beforeStreamCall` failure behavior, and `onStreamChunk -> null` filtering. Please add a deterministic unit test around `InterceptorChain.applyStreamingInterceptors(...)`; otherwise the two edge cases above can regress without CI noticing.

**Review by Aias00 — CHANGES_REQUESTED** [review]
Thanks for adding a chunk-level interceptor extension point. I found two blocking issues in `applyStreamingInterceptors(...)` and one missing test case before this is safe to merge:

- the per-subscription flow still reuses the same mutable `ModelRequest` instance across subscribers
- exceptions from `beforeStreamCall(...)` bypass `onStreamError(...)`
- the new tests only cover the live happy path and do not pin down the new retry/multi-subscribe semantics

Requesting changes so these lifecycle guarantees are nailed down before merge.

**Review by Jackie9991 — COMMENTED** [review]
Thanks for the careful review @Aias00 — all three are addressed in 9c4c6c9:

**1. Per-subscription `ModelRequest` isolation.** The defer block now starts each subscription from a defensive copy of the request: `ModelRequest.builder(src)` (which already shallow-copies the small collection fields) plus an explicit `new ArrayList<>(messages)` and `options.copy()` so the two remaining shared-reference fields no longer leak. An interceptor that does `request.getContext().put(...)` or `request.getOptions().setTemperature(...)` now mutates only its subscription's copy.

**2. `beforeStreamCall` exceptions surface through `onStreamError`.** The threading phase is wrapped in try/catch: on throw we fan out `onStreamError(originalThrowable, scopedRequest)` to every interceptor (suppressing any callback-thrown errors so they don't mask the original), then return `Flux.error(originalThrowable)` so the subscriber observes the same exception. Documented the lifecycle guarantee in the chain Javadoc.

**3. Deterministic unit tests.** Added `InterceptorChainStreamingTest` (no DashScope dependency, runs in vanilla CI) covering exactly the four risky semantics:
- multi-subscribe aggregator isolation (second subscription sees `"ab"`, not `"abab"`)
- `onStreamChunk -> null` drops the chunk from downstream **and** from the aggregated message
- `beforeStreamCall` throw → all interceptors' `onStreamError` invoked + original throwable propagated downstream
- in-place context mutation in `beforeStreamCall` does not leak across subscriptions

PTAL when you have time.

**Review by Aias00 — APPROVED** [review]
LGTM
